Implicit import fixes

hnrklssn opened the pull request:
This fixes some bugs related to implicitly inherited imports that block enabling safe interop wrappers by default.

rdar://161795145
rdar://161795429
rdar://161795673
rdar://161795793

hnrklssn committed:
[MacrosOnImports] Don't import !swift modules

Inheriting a clang import marked `requires: !swift` will always result
in an error. This skips such imports, which *may* result in name lookup
errors instead, but also may not, depending on the module.

rdar://161795145

hnrklssn committed:
[ImportResolution] Only disallow explicit `std` imports (allow implicit)

When expanding a Swift macro in a clang module where the original clang
module imported a submodule in a C++ standard library module other than
`std`, e.g. a submodule to `std_core`, this would result in an error.
This is because `std_core.math.abs` would be imported as
`CxxStdlib.math.abs`, which would later be translated back as
`std.math.abs` which doesn’t exist.

This changes the mapping to only map `std` to `CxxStdlib`. To prevent
errors when importing modules starting with `std_`, this error is moved
from the late-stage module import to the earlier processing of
`ImportDecl`s. This results in these module names still being forbidden
in explicit imports (i.e. naming them in source code), while still being
allowed in implicit imports inherited from clang modules.

This also fixes a fix-it bug where only the first 3 characters would be
selected for replacing with `CxxStdlib` when importing `std_core`.

This also fixes a diagnostic bug where aliased modules would refer to
the module name in the source code rather than the real module name, and
adds a note clarifying the situation.

rdar://161795429
rdar://161795673
rdar://161795793

hnrklssn commented:
@swift-ci please smoke test

hnrklssn committed:
remove unnecessary c++ interop

hnrklssn commented:
@swift-ci please smoke test

Xazax-hun commented on lib/Sema/ImportResolution.cpp:
```diff
@@ -359,6 +359,26 @@ static bool isSubmodule(const ModuleDecl* M) {
+  if (front == "std" || front.starts_with("std_")) {
+    SmallString<64> modulePathStr;
+    path.getString(modulePathStr);
+    auto diagKind = (ctx.LangOpts.DebuggerSupport || true) ? diag::sema_no_import_repl
```
Is the `|| true` intended?

hnrklssn replied:
Oh wow! Nope, forgot to clean that up.

egorzhdan commented on lib/ClangImporter/ClangImporter.cpp:
```diff
@@ -8808,7 +8804,7 @@ bool importer::isCxxStdModule(StringRef moduleName, bool IsSystem) {
   ImportPath::Builder builder;
   while (M) {
-    if (!M->isSubModule() && isCxxStdModule(M))
+    if (!M->isSubModule() && M->Name == "std")
```
Should this still accept `std_core`, which acts as a top-level module in recent libc++?

hnrklssn committed:
fix non-fatal import error

hnrklssn commented:
@swift-ci please smoke test

hnrklssn commented:
@swift-ci please smoke test

hnrklssn committed:
only run transitive-std-core-import.swift on new libc++

hnrklssn commented:
@swift-ci please smoke test

hnrklssn committed:
[DiagnosticVerifier] Add -verify-ignore-macro-note

The _SwiftifyImport macro is emitted into an unnamed buffer and then
parsed, pretending it was in the header all along. This makes it hard to
add `expected-note` comments for `diag::in_macro_expansion` when they
point here. That's okay, because the macro expansion has already been
pointed out by `expected-expansion` directives. But
-verify-ignore-unrelated is too blunt of a tool, so this adds
-verify-ignore-macro-note to ignore these specific diagnostics.

hnrklssn committed:
fix test on Windows

hnrklssn committed:
don't run %empty-directory twice

hnrklssn committed:
Only inherit C++ modules when C++ interop enabled

Imports of modules marked `requires: cplusplus` should only be inherited
when C++ interop is enabled. Similarly, imports of modules marked
`requires: !cplusplus` should only be inherited when C++ interop is
disabled.

hnrklssn commented:
Closing this as it has been split into https://github.com/swiftlang/swift/pull/84871, https://github.com/swiftlang/swift/pull/84870 and https://github.com/swiftlang/swift/pull/84896
